core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -58,21 +66,19 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
       // case where the converted field is non-null is when a map or list is
       // selected by lower IDs.
       if (selectedIds.contains(fieldId)) {
-        filteredFields.add(copyField(field, field.schema()));
+        filteredFields.add(copyField(field, field.schema(), fieldId));
       } else if (fieldSchema != null) {
-        hasChange = true;
-        filteredFields.add(copyField(field, fieldSchema));
+        filteredFields.add(copyField(field, fieldSchema, fieldId));
       }
     }
 
-    if (hasChange) {
+    if (filteredFields.size() > 0) {
```

Comment:
This shouldn't remove the `hasChange` flag. The purpose of `hasChange` is to signal when this is using the projection of a field. That way, if there are no known changed fields and all fields were projected, we use the original record. That's better for caching.

Since you now need to detect the case where the field ID is missing so that that record gets recreated, you should add a case above that detects when it isn't there and sets `hasChange = true`.

Reply:
Makes sense. I've updated the patch